File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameter = 0;

    int diameterOfBinaryTree(TreeNode* root) {
        // length is number of edges between any two nodes
        // find longest length

        // intuition 1: level traversal, suppose we know the result of child nodes
        // intuition 2: height

        heightOfNode(root);
        return diameter;
    }

    int heightOfNode(TreeNode* root) {
        if (!root) return 0;

        int left = heightOfNode(root->left);
        int right = heightOfNode(root->right);

        diameter = max(diameter, left + right);

        return max(left, right) + 1;
    }
};

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        # in-order traversal - recursive dfs
        # current node counted after left traversal
        # after find the 'kth' return any left recursion

        res: Treenode

        def dfs(curr):
            nonlocal res, k
            
            if not curr: return 

            dfs(curr.left) # delve left

            if k == 0: return 
            k -= 1
            if k == 0: 
                res = curr.val
                return
            
            dfs(curr.right) # delve right

        dfs(root)

        return res



